libr/include/r_util/r_assert.h

```diff
@@ -0,0 +1,97 @@
+#ifndef R_ASSERT_H
+#define R_ASSERT_H
+
+#if defined (__GNUC__) && defined (__cplusplus)
+#define R_STRFUNC     ((const char*) (__PRETTY_FUNCTION__))
+#elif defined (__STDC_VERSION__) && __STDC_VERSION__ >= 199901L
+#define R_STRFUNC     ((const char*) (__func__))
+#elif defined (__GNUC__) || (defined(_MSC_VER) && (_MSC_VER > 1300))
+#define R_STRFUNC     ((const char*) (__FUNCTION__))
+#else
+#define R_STRFUNC     ((const char*) ("???"))
+#endif
+
+typedef enum r_log_level {
+	R_LOG_DEBUG = 0,
+	R_LOG_INFO,
+	R_LOG_WARNING,
+	R_LOG_ERROR,
+	R_LOG_CRITICAL
+} RLogLevel;
+
+R_API void r_log(RLogLevel level, const char *fmt, ...);
+
+#define r_warn_if_reached()						\
+	do {								\
+		r_log (R_LOG_WARNING, "(%s:%d):%s%s code should not be reached\n", \
+		       __FILE__, __LINE__, R_STRFUNC, R_STRFUNC[0] ? ":" : "");	\
+	} while (0)
+
+#define r_warn_if_fail(expr)					\
+	do {								\
+		if (!(expr)) {						\
+			r_log (R_LOG_WARNING, "WARNING (%s:%d):%s%s runtime check failed: (%s)\n", \
+			       __FILE__, __LINE__, R_STRFUNC, , R_STRFUNC[0] ? ":" : "", #expr); \
+		}							\
+	} while (0)
+
+#ifdef R_DISABLE_CHECKS
```

Comment:
what about using `#if` instead of `#ifdef`?

We can then have different levels of check behaviours, as well as be able to use =0

Reply:
im still thinking if having more behaviours for this and have an option to enable real asserts in place.

The variable name should be different. and we must decide which configure option use to enable or disable this and which should be the default one.

Reply:
`R_CHECKS_LEVEL`

* 0: disabled, does nothing, not even return, so, crashes are expected to happen, but those may show us real errors that must be fixed, obviously not handy for end users, but good for us as developers.
* 1: enabled, silently return (this is the current behaviour in r2land)
* 2: logging enabled, shows a warning message to stderr and returns.
* 3: ??

in case of debugging, we can always set a breakpoint in `r_log` (or whatever api we decide to use to break in there, or we can add another variable to specify which log level messages should call r_sys_breakpoint().

This way we can get different types of builds. but we need to decide which ones should be set by default.

Imho, the less painful one would be to keep the current behaviour (aka R_CHECKS_LEVEL=1), but we as developers can use 0 or 2 for debugging purposes.

Eventually we can enable them in travis to find bugs, and the logs should be verbose enough to be able to find the problem. So not sure if we should also include an `r_sys_backtrace()` in this output. that's why i think the `r_log` api is a bad idea here because this will be called a lot of times and we must reduce the logic code that is copied there into an external function that does the rest of the job.